Quest Details View Implemented

jmgoldb opened the pull request:
This one seemed a little bit sprawling behind the scenes, though all the functionality is working exactly as intended.  Can I get a code review, please?

jmgoldb committed:
Merge branch 'questHandling'

jmgoldb committed:
Merge commit 'cbfbf28e2c3ff87614e36ed0a7ce2d8bded4d257'

jmgoldb committed:
Completed Quest Details View, including moving from quest to monster  Up next is hooking up monsters back to their quests in the other direction, followed by quest drops

jmgoldb commented:
@vokal/ios-developers Specifically, I'm not sure if it's a good idea to be using nested switch statements the way that I am.

brockboland commented:
Let's have a look-see here

brockboland commented on MH4U Dex/QuestContainerViewController.m:
```diff
@@ -0,0 +1,61 @@
+
+- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
+{
+    if ([segue.identifier isEqualToString:@"embedQuestDetails"]) {
```
You can get away with it here, but it's generally a good to put storyboard identifies in constants; structs are often used for this.

jmgoldb replied:
You mentioned I should put the constants in a struct, is there a particular reason for that?

brockboland replied:
Just so they're grouped. You might access this one as `SegueIdentifiers.EmbeddedQuestDetails` and the other as `SegueIdentifiers.ShowQuestContainer`. I think that most of our projects use this strategy, and put the struct in the constants file.

jmgoldb replied:
I'll take a look at how nextTier does it

brockboland commented on MH4U Dex/QuestDetailsTableViewController.m:
```diff
@@ -0,0 +1,189 @@
+
+- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
+{
+    return self.questMonsters.count ? 2 : 1;
```
Could you include a comment explaining what these numbers mean?

jmgoldb replied:
Sure

jmgoldb committed:
Added a comment explaining the meaning of why QuestDetailsTable has a tertiary operator in determining number of sections

brockboland commented on MH4U Dex/QuestDetailsTableViewController.m:
```diff
@@ -0,0 +1,189 @@
+
+- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
+{
+    UIStoryboard *storyBoard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
```
I'd also recommend a constant for the storyboard name

brockboland commented on MH4U Dex/QuestDetailsTableViewController.m:
```diff
@@ -0,0 +1,189 @@
+            break;
+        case 1:
+            cell.userInteractionEnabled = YES;
+            monster = self.questMonsters[indexPath.row];
```
You could be using your `monsterForIndexPath:` helper here

brockboland commented on MH4U Dex/QuestListTableViewController.m:
```diff
@@ -32,6 +34,17 @@ - (void)viewDidLoad
+- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
+{
+    QuestContainerViewController *questVC = (QuestContainerViewController *)segue.destinationViewController;
+    if ([segue.identifier isEqualToString:@"showQuestContainerView"]) {
```
Same thing here about constants for segue identifier

brockboland commented:
A few suggestions.
1. In QuestDetailsTableViewController, I'd use an enum for the sections, just like you did in QuestContainerViewController. In this case, the count doesn't need to be part of the enum, since the number of sections is dependent on the contents of the `questMonsters` property. But, having a bunch of 0's and 1's in the various table view datasource/delegate methods isn't very clear.
2. You asked about the nested switch statements: yeah, that method is kind of a beast now. I would split that out into two helper methods that create and configure the two types of cells, and then call one of those depending on the section that the cell will go in.
3. The cell config for the first section is kinda funky. At the very least, I'd add comments explaining that (for example) the caravan/hall and other values are bool numbers, and so on.
4. The `QuestDetailEntries` and constant strings seem pretty verbose, but I'm not sure there's a much better way to handle it. I think I would combine them into an array, like this:

``` objc
NSArray *questDetailLabels = @[
                               @"Caravan/Hall",
                               @"Stars",
                               @"Objective",
                               @"Sub Objective",
                               @"Quest Type",
                               @"Key Quest",
                               @"Urgent Quest",
                               @"Unstable Quest",
                               @"Quest Fee",
                               @"Zenny Reward",
                               @"HRP Reward",
                               ];                           
```

Then, when building the cell:

``` objc
NSString *rowKey = questDetailLabels[indexPath.row];
```

There would still be some funky handling for the cases that use a boolean number, but it might help clean things up a bit.

jmgoldb committed:
Moved some storyboard identifiers into structs in Constants.h

Combined QuestDetailEntries and several static Strings into an NSArray

Split QuestDetailsTableViewController's cell generation into two separate methods to avoid unnecessary switch statements.

jmgoldb committed:
Forgot to include these files in the commit!

jmgoldb commented:
Alrighty, I think that should take care of everything, but I feel weird calling on the indices of questDetailLabels directly.  Can you give me your opinion?

brockboland commented:
Looks a lot better. The `questDetailLabels` indices…yeah, I'm not sure how to handle that best. Maybe it would make sense to keep that big `switch` statement in `detailCellForIndexPath:` after all, and lose `questDetailsDict` completely.

jmgoldb commented:
I dunno, I think its far more readable without the giant switch statement.  And if don't use questDetailsDict, then I have to make things even more sprawling by manually assigning each property of the quest model object.

brockboland commented:
True. I'm fine with this as it is, but keep it at the back of your mind in case you come up with a better option.

Good to pull

jmgoldb commented:
Will do.  Thanks

jmgoldb commented:
@chillpop @seanwolter @designatednerd Can I get another merge, please?

seanwolter commented:
:suspect:

seanwolter commented:
That's the "suspect" face. 
I guess if Brock sez it's good to pull...

jmgoldb commented:
Isn't that the doomguy face?

seanwolter commented:
@jmgoldb They're all here for you, my friend: http://www.emoji-cheat-sheet.com

scottferg commented:
:neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard::neckbeard:
